### PR TITLE
spirv-fuzz: Report fresh ids in transformations

### LINKED ISSUE #3851: spirv-fuzz: Transformations should report the fresh ids that they use
For overflow ids to work properly, we need to be able to determine, from a sequence of transformations, those fresh ids that might be used when replaying the sequence.  To allow this, Transformation should have a virtual method, GetFreshIds(), that every implementing class should use to specify which fresh ids each transformation uses.

### PR DESCRIPTION
**Pull request by afd**
Adds a virtual method, GetFreshIds(), to Transformation.  Every
transformation uses this to indicate which ids in its protobuf message
are fresh ids.  This means that when replaying a sequence of
transformations the replayer can obtain a smallest id that is not in
use by the module already and that will not be used by any
transformation by necessity.  Ids greater than or equal to this id
can be used as overflow ids.

Fixes #3851.

### CODE REVIEW

**Review by Vasniktel — COMMENTED** [review]
Looks good. I've left a question about NullOverflowIdSource.

**Review by paulthomson — APPROVED** [review]
LGTM! I did not check every test nor every transformation, but I think I have checked pretty much everything else.

File: source/fuzz/replayer.cpp [hunk 1]
```diff
@@ -90,13 +90,24 @@ Replayer::ReplayerResult Replayer::Run() {
...
   std::unique_ptr<TransformationContext> transformation_context =
-      first_overflow_id_ == 0
-          ? MakeUnique<TransformationContext>(
-                MakeUnique<FactManager>(ir_context.get()), validator_options_)
-          : MakeUnique<TransformationContext>(
-                MakeUnique<FactManager>(ir_context.get()), validator_options_,
-                MakeUnique<CounterOverflowIdSource>(first_overflow_id_));
+      MakeUnique<TransformationContext>(
+          MakeUnique<FactManager>(ir_context.get()), validator_options_,
+          MakeUnique<CounterOverflowIdSource>(first_overflow_id));
```
[the remark this answers is not in the document]
Comment (paulthomson) [reply source/fuzz/replayer.cpp:110]: I think we still use it when fuzzing. 

https://github.com/KhronosGroup/SPIRV-Tools/blob/9edeeafdb64213453a0b33a8acb33b8418b40d46/source/fuzz/transformation_context.h#L34

[the remark this answers is not in the document]
Comment (paulthomson) [reply source/fuzz/replayer.cpp:110]: Updated link

[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/replayer.cpp:110]: Thanks both. @paulthomson is correct that we do still need NullOverflowIdSource.

**Pull request merged**